klippy/extras/probe_eddy_current.py

```diff
@@ -345,22 +451,43 @@ def home_wait(self, home_end_time):
         return trigger_time
     # Probe session interface
     def start_probe_session(self, gcmd):
-        self._gather = EddyGatherSamples(self._printer, self._sensor_helper,
-                                         self._calibration, self._z_offset)
+        method = gcmd.get('METHOD', 'automatic').lower() if gcmd else None
+        self._is_tap = method == 'tap'
+        if self._is_tap:
+            self._gather = EddyGatherTapSamples(self._printer,
+                                                self._sensor_helper,
+                                                self.tap_max_decel)
+        else:
+            self._gather = EddyGatherSamples(self._printer,
+                                             self._sensor_helper,
+                                             self._calibration, self._z_offset)
         return self
     def run_probe(self, gcmd):
         toolhead = self._printer.lookup_object('toolhead')
         pos = toolhead.get_position()
+        # Safety duct tape
+        probe_params = self._param_helper.get_probe_params(gcmd)
+        if pos[2] < self._z_offset:
+            lift_speed = probe_params['lift_speed']
+            retract_dist = probe_params['sample_retract_dist']
+            pos[2] = self._z_offset + retract_dist
+            toolhead.move(pos, lift_speed)
```

Comment:
I don't think this is a good idea.  If it is not valid to probe when the Z is below z_offset then the code should raise an error.  I think it's a really bad idea to "try to help the user by doing something they didn't ask for" - it causes very frustrating experiences for users as they often perceive the printer as doing "weird things" and the alternate implementations frequently produce bad results.